File: src/main/java/edu/cornell/cs/apl/viaduct/errors/NameClashError.java
package edu.cornell.cs.apl.viaduct.errors;

import edu.cornell.cs.apl.viaduct.imp.ast.Name;
import edu.cornell.cs.apl.viaduct.imp.parsing.HasLocation;
import edu.cornell.cs.apl.viaduct.imp.parsing.SourceRange;
import java.io.PrintStream;

/** A (variable, host, process etc.) name declared multiple times. */
public class NameClashError extends CompilationError {
  private final Name name;
  private final SourceRange firstDeclaration;
  private final SourceRange secondDeclaration;

  /**
   * Create an error description given two definition sites of the same name.
   *
   * @param firstDeclaration original location the name is declared
   * @param secondDeclaration current declaration
   */
  public <N extends HasLocation & Name> NameClashError(
      HasLocation firstDeclaration, N secondDeclaration) {
    this.name = secondDeclaration;
    this.firstDeclaration = firstDeclaration.getSourceLocation();
    this.secondDeclaration = secondDeclaration.getSourceLocation();

    assert this.firstDeclaration.getSourcePath().equals(this.secondDeclaration.getSourcePath());
  }

  @Override
  protected String getCategory() {
    return "Name Clash";
  }

  @Override
  protected String getSource() {
    return firstDeclaration.getSourcePath();
  }

  @Override
  public void print(PrintStream output) {
    super.print(output);

    output.print("This file has multiple ");
    name.print(output);
    output.println(" declarations. One here:");

    output.println();
    firstDeclaration.showInSource(output);

    output.println("And another one here:");

    output.println();
    secondDeclaration.showInSource(output);
  }
}

File: src/main/kotlin/edu/cornell/cs/apl/viaduct/errorskotlin/NameClashError.kt
package edu.cornell.cs.apl.viaduct.errorskotlin

import edu.cornell.cs.apl.viaduct.syntax.Name
import edu.cornell.cs.apl.viaduct.syntax.SourceLocation
import java.io.PrintStream

/** Thrown when a [Name] has multiple declarations. */
class NameClashError(
    private val name: Name,
    private val firstDeclaration: SourceLocation,
    private val secondDeclaration: SourceLocation
) : CompilationError() {
    init {
        require(this.firstDeclaration.sourcePath == this.secondDeclaration.sourcePath)
    }

    override val category: String
        get() = "Name Clash"

    override val source: String
        get() = firstDeclaration.sourcePath

    override fun print(output: PrintStream) {
        super.print(output)

        output.print("This file has multiple declarations of ")
        output.print(name.name) // TODO: Printer.run(name, output)

        output.println(". One here:")

        output.println()
        firstDeclaration.showInSource(output)

        output.println("And another one here:")

        output.println()
        secondDeclaration.showInSource(output)
    }
}
